# Day17_RandomForest_Hyperparameters


In [ ]:
import pandas as pd


In [ ]:
df=pd.read_csv("/content/train.csv")

**Cleaning**

In [ ]:
df=df.drop(["PassengerId","Name","Ticket"],axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


In [ ]:
df["Age"]=df["Age"].fillna(df["Age"].median())

In [ ]:
df["Embarked"]=df["Embarked"].fillna(df["Embarked"].mode())

In [ ]:
df["Sex"]=df["Sex"].map({"male":0,"female":1})

In [ ]:
df["Sex"].value_counts(dropna=False)

,count
Sex,
0,577
1,314


In [ ]:
df=pd.get_dummies(df,columns=["Embarked"],drop_first=True)

In [ ]:
df.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin',
       'Embarked_Q', 'Embarked_S'],
      dtype='object')

In [ ]:
df.isnull().sum()
df.dtypes

,0
Survived,int64
Pclass,int64
Sex,int64
Age,float64
SibSp,int64
Parch,int64
Fare,float64
Cabin,object
Embarked_Q,bool
Embarked_S,bool


In [ ]:
#Logistic Regression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [ ]:
y=df["Survived"]
df=df.drop(["Survived"],axis=1)
x=df

In [ ]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Pclass      891 non-null    int64  
 1   Sex         891 non-null    int64  
 2   Age         891 non-null    float64
 3   SibSp       891 non-null    int64  
 4   Parch       891 non-null    int64  
 5   Fare        891 non-null    float64
 6   Cabin       204 non-null    object 
 7   Embarked_Q  891 non-null    bool   
 8   Embarked_S  891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(1)
memory usage: 50.6+ KB


In [ ]:
x = x.drop("Cabin",axis=1)
from sklearn.linear_model import LogisticRegression
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)
lr=LogisticRegression(max_iter=1000)
lr.fit(x_train,y_train)
train_acc_lr=accuracy_score(y_train,lr.predict(x_train))
test_acc_lr=accuracy_score(y_test,lr.predict(x_test))
train_acc_lr,test_acc_lr

(0.800561797752809, 0.8100558659217877)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
dt=DecisionTreeClassifier(max_depth=3,random_state=42)
dt.fit(x_train,y_train)
train_acc_dt=accuracy_score(y_train,dt.predict(x_train))
test_acc_dt=accuracy_score(y_test,dt.predict(x_test))
train_acc_dt,test_acc_dt

(0.8342696629213483, 0.7988826815642458)

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(random_state=42)
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.9789325842696629, 0.7988826815642458)

Logistic Regression
Train:0.800561797752809
Test:0.8100558659217877

Decision Tree
Train:0.8342696629213483
Test:0.7988826815642458

Random Forest
Train:0.9789325842696629
Test:0.7988826815642458

Which model generalizes best?
Decision Tree
Which model shows signs of overfitting?
Random Forest

In [ ]:
rf=RandomForestClassifier(n_estimators=50,random_state=42)#n_estimators is number of tress
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.9789325842696629, 0.8044692737430168)

#When n_estimators is changed to 50
What happens to train accuracy
train accuracy ~0.98
What happens to test accuracy
test accuracy ~0.80
here model overfits
Increasing n_estimators does NOT increase overfitting

Overfitting is mainly due to tree depth & flexibility

Here, trees are still deep → ensemble memorizes training data

So say this instead:

Even with more trees, overfitting remains because individual trees are too complex.

In [ ]:
rf=RandomForestClassifier(max_depth=5,random_state=42)
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.8581460674157303, 0.8156424581005587)

#max_depth = 5
What happens to train accuracy
0.8581460674157303
What happens to test accuracy
0.8156424581005587
here model doesn't overfits or underfit's

In [ ]:
rf=RandomForestClassifier(random_state=42)
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.9789325842696629, 0.7988826815642458)

#max_depth = None
What happens to train accuracy
0.9789325842696629
What happens to test accuracy
0.7988826815642458
Here,model overfits

Does limiting depth help?
Yeah,it helps the model not to overfits
Why Random Forest is more stable than a single tree?
because it helps to reduce the model overfits

1)Why is Random Forest less prone to overfitting than Decision Tree?
Choose Logistic Regression when:

relationship is roughly linear

interpretability is important

dataset is small

you want stable, explainable coefficients

2)What does “ensemble” mean?
It combines multiple models to create single model ,improving predictive accuracy and reliability and reduces errors,bias,overfitting

3)When would you choose Logistic Regression over Random Forest?
when all features are equally important to make desicion for Logestic Regression
when not all features ar equally important Random Forest

Day 17 Summary

Learned what n_estimators means in Random Forest

Compared Logistic Regression, Decision Tree, and Random Forest performance

Observed overfitting using train vs test accuracy

Understood how increasing trees and limiting depth affects overfitting

Learned why Random Forest is more stable than a single Decision Tree

# Day18_ModelTuning

max_depth → controls complexity (overfitting)

n_estimators → number of trees (stability)

min_samples_split -> defines the minimum number of samples required to split an internal node. It acts as a regularization parameter to control the complexity of individual trees within the forest and prevent overfitting.

min_samples_leaf -> the minimum number of samples that must be present in a leaf node (terminal node) after a split. Its primary purpose is to control the complexity of individual decision trees within the forest and prevent overfitting.

In [ ]:
rf=RandomForestClassifier(n_estimators=100,max_depth=2,random_state=42)
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.8061797752808989, 0.770949720670391)

train accuracy
0.8061797752808989
test accuracy
0.770949720670391
Did gap reduce or increase?
decreased

In [ ]:
rf=RandomForestClassifier(n_estimators=10,max_depth=5,random_state=42)
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.8637640449438202, 0.7988826815642458)

train accuracy
0.8637640449438202
test accuracy
0.7988826815642458
Did gap reduce or increase?
decreased

In [ ]:
rf=RandomForestClassifier(n_estimators=100,max_depth=5,random_state=12)
rf.fit(x_train,y_train)
train_acc_rf=accuracy_score(y_train,rf.predict(x_train))
test_acc_rf=accuracy_score(y_test,rf.predict(x_test))
train_acc_rf,test_acc_rf

(0.8581460674157303, 0.8100558659217877)

train accuracy
0.8581460674157303
test accuracy
0.8100558659217877
Did gap reduce or increase?
decreased

Why increasing n_estimators does not always fix overfitting?
its not true that if you increase n_estimators it will fix overfitting. it might become worse
Which parameter helped more in your case:
n_estimators or max_depth?
max_depth
When would you stop tuning a model?
I would stop tuning when improvements in validation/test performance become marginal and the train test gap is small and stable.

Day 18 Summary

Learned how hyperparameters control model complexity and overfitting

Understood the role of max_depth, n_estimators, min_samples_split, and min_samples_leaf

Experimented with Random Forest hyperparameter tuning

Observed how limiting tree depth reduced overfitting more effectively than increasing the number of trees

Learned that increasing n_estimators improves stability but does not always improve generalization

Understood when to stop tuning a model by monitoring train–test gap and test accuracy stability

# DAY 19-Model Comparison & Selection

**Logistic Regression**
Train accuracy
0.800561797752809

Test accuracy
0.8100558659217877
Overfitting? (Yes / No)
No
Stability (High / Medium / Low)
Medium

**Decision Tree (depth-limited)**
Train accuracy
0.8342696629213483
Test accuracy
0.7988826815642458
Overfitting? (Yes / No)
No
Stability (High / Medium / Low)
Low

**Random Forest (tuned)**
Train accuracy
0.8581460674157303

Test accuracy
0.8156424581005587

Overfitting? (Yes / No)
No
Stability (High / Medium / Low)
Medium

Which model would you choose for deployment? Why?
Random Forest


Which model overfits the most? Why?
Random Forest (when not tuned), because it memorizes training data with deep trees.

Which model is easiest to explain to a non-technical person?
Logistic Regression

Accuracy vs Interpretability — which did you prioritize and why?
Interpretability was prioritized because simpler models are easier to trust, explain, and debug in real-world decision making.

“Even though Random Forest had higher train accuracy, I chose _Logestic regression__ because _it is easy to interpret__.”

Day 19 Summary – Model Comparison & Selection

Compared Logistic Regression, Decision Tree, and Random Forest using train/test accuracy.

Observed that high training accuracy alone does not indicate a good model.

Identified overfitting behavior using train–test accuracy gaps.

Found Logistic Regression to be stable and interpretable.

Observed Decision Trees are sensitive to data splits but improve with depth control.

Learned that Random Forest performs well but can overfit if not properly tuned.

Chose Logistic Regression for deployment due to better interpretability and reliable generalization.

Understood the trade-off between accuracy and interpretability in real-world applications.

## Day 20-how to evaluate a model beyond accuracy and how confident the model

In [ ]:
y_prob = lr.predict_proba(x_test)
y_prob[:5]

array([[0.88866292, 0.11133708],
       [0.77838766, 0.22161234],
       [0.86562027, 0.13437973],
       [0.1142991 , 0.8857009 ],
       [0.26153537, 0.73846463]])

What do these two columns represent?
The first column represents the probability of class 0 (Not Survived) and the second column represents the probability of class 1 (Survived) for each test sample.

In [ ]:
from sklearn.metrics import roc_auc_score
roc_auc=roc_auc_score(y_test,y_prob[:,1])
roc_auc

np.float64(0.883011583011583)

Is ROC-AUC better or worse than accuracy here?and why?
ROC–AUC and accuracy are NOT comparable numbers

ROC–AUC measures ranking quality

Accuracy depends on chosen threshold (default = 0.5)

ROC–AUC is not better or worse than accuracy; it measures how well the model separates classes across all thresholds, while accuracy depends on a fixed threshold.

In [ ]:
import numpy as np
y_pred_custom=(y_prob[:,1]>0.3).astype(int)
y_pred_custom

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1,
       1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1,
       1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0,
       1, 1, 1])

In [ ]:
y_pred_custom=(y_prob[:,1]>0.5).astype(int)
y_pred_custom

array([0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 1, 1])

Lower threshold (0.3) → predicts more survivors

Higher threshold (0.5) → stricter, fewer positives

You don’t need accuracy here, just understanding behavior.
What changed?
there are few differences when thresholds changes

Which errors increased/decreased?
Changing the threshold changes how confident the model must be before predicting survival, which affects false positives and false negatives.

Day 20 Summary

Learned how to interpret prediction probabilities using Logistic Regression

Understood what class probabilities represent in predict_proba

Learned ROC–AUC as a threshold-independent evaluation metric

Explored how changing decision thresholds affects predictions

Understood why accuracy alone is not sufficient to evaluate model confidence

# Day 21: Precision–Recall & Threshold Trade-offs

**Task 1 Precision & Recall intuition **

Precision = Of all predicted survived, how many actually survived?

Recall = Of all people who survived, how many did the model catch?

If we want to avoid false positives, do we increase precision or recall?
Precision
If we want to avoid missing true survivors, do we increase precision or recall?
Recall


In [ ]:
y_pred=lr.predict(x_test)

In [ ]:
from sklearn.metrics import precision_score,recall_score
precision=precision_score(y_test,y_pred)
recall=recall_score(y_test,y_pred)
precision,recall

(0.7857142857142857, 0.7432432432432432)

Is your model more precision-oriented or recall-oriented?
Precision

In [ ]:
import numpy as np
y_pred_03=(y_prob[:,1]>0.3).astype(int)
y_pred_05=(y_prob[:,1]>0.5).astype(int)
y_pred_07=(y_prob[:,1]>0.7).astype(int)

In [ ]:
for t,yp in [(0.3,y_pred_03),(0.5,y_pred_05),(0.7,y_pred_07)]:
  print(t,precision_score(y_test,yp),recall_score(y_test,yp))

0.3 0.6914893617021277 0.8783783783783784
0.5 0.7857142857142857 0.7432432432432432
0.7 0.9318181818181818 0.5540540540540541


When threshold ↓, what happens to recall?
recall increaces

When threshold ↑, what happens to precision?
precision increases

Why is there no single best threshold?
it changes based on the dataset and models

There is no single best threshold because the optimal trade-off between precision and recall depends on the problem’s cost of false positives vs false negatives, the dataset distribution, and the business objective.

Day 21 Summary

Write 4 bullets:

Learned difference between precision and recall

Understood threshold trade-off

Observed how probabilities affect predictions

Realized why accuracy alone is misleading